Let's keep paragraph separators.

searchivarius opened the pull request:
Thank you for this nice text extractor! It extracts and cleans up text really well. It is also useful that you can get what separate paragraphs are, b/c there is a newline. However, in rare cases a newline occurs only because a user entered it while editing a Wikipedia article.

Check, for example, [this paragraph ](https://en.wikipedia.org/w/index.php?title=Alain_Connes&action=edit&section=1&editintro=Template:BLP_editintro). There is an editor-introduced newline after the phrase ``which culminated in the``. This newline is then propagated to the extracted text. 

So, this prevents me from correctly splitting the text into paragraphs sometimes. As a simplest fix, I suggest keeping one empty line as an indicator for a paragraph separator.

Thanks!

attardi commented:
I don't understand how your change fixes the problem.
There are no multiple empty lines in your example.

searchivarius commented:
Hi @attardi the fix retains an empty line that indicates a *new* paragraph. Currently, you delete it. This is a problem, because a single newline doesn't always mean a start of a paragraph. An example of when this is not the case is given above. You have a newline in the *middle* of the paragraph.

attardi commented:
Look, your fix does not handle the problem. The code splits the text by newlines:

    for line in text.split('\n'):

hence in the example you would get two lines, none of which is empty. 
The test
    if line.strip() == '':
will not be true and hence it would have no effect.

BTW:
    if line.strip() == '':
works the same as:
    if not line:
since there are no lines with just spaces to be stripped.

searchivarius commented:
Ok, it may be clearer with an example. The piece produced by the current code vs the piece produced by the modified one. Current code:
```
Alain Connes studies operator algebras. In his early work on von Neumann algebras in the 1970s, he succeeded in obtaining the almost complete classification of injective factors. Following this he made contributions in operator K-theory and index theory, which culminated in the 
Baum–Connes conjecture. He also introduced cyclic cohomology in the early 1980s as a first step in the study of noncommutative differential geometry. He was a member of Bourbaki.
Connes has applied his work in areas of mathematics and theoretical physics, including number theory, differential geometry and particle physics.
```
Updated one:
```
Alain Connes studies operator algebras. In his early work on von Neumann algebras in the 1970s, he succeeded in obtaining the almost complete classification of injective factors. Following this he made contributions in operator K-theory and index theory, which culminated in the 
Baum–Connes conjecture. He also introduced cyclic cohomology in the early 1980s as a first step in the study of noncommutative differential geometry. He was a member of Bourbaki.

Connes has applied his work in areas of mathematics and theoretical physics, including number theory, differential geometry and particle physics.
```
I don't remove the new line after ``which culminated in the `` (though I could have), it just retain empty pargraph lines. Your code **does produce paragraph-separating lines**, but they are later removed.

Again, the code of the function **compact** has the line:
```
if not line:
  continue
```
This is where you get at least one empty line, but currently it is droppped.

Anyways, never mind.

attardi commented:
I understand what you want.
A simpler code is this:

    if not line:
       # collapse multiple empty lines
       if page and page[-1]:
          page.append('')
       continue
